### PR TITLE
Create mapping order

### PR DESCRIPTION
**Pull request by carlvitzthum**
This branch contains changes to order indexing triggered by create mapping by item type. It also contains a bunch of random changes and refactors (sorry). Below is a list of stuff changed:
- **create_mapping**: Refactored/simplified a bunch of outdated code. Previously a set, `uuids_to_index` is now a dict of sets of uuids to index, keyed by item type. Additionally, `create_mapping.run` accepts an `item_order` parameter that will order the uuids from `uuids_to_index` by the index of the item type in that parameter. This means if an item is at the top of the list, it will more-or-less get indexed first (it's not that simple because of multi-processing indexing + SQS being distributed, but you get the idea)
- **fourfront_utils**: I didn't like the name of this module so I removed it. All functions have been moved to `util.py`, which is now rearranged a bit as well
- **indexer_utils**: docstrings and refactored `get_uuids_for_types` to make simpler
- **resources. AbstractCollection**: added the `iter_no_subtypes` method, which is a generator for all items in the collection but NOT items in the subtype collections. The regular __iter__ generator returns all the items (e.g. `for item in collection: ...`)
- **resources.Item**: since we are referencing these fields explicitly in snovault, I added `principals_allowed` and `display_title` calculated properties to the base Item
- **default embeds**: `@type` is now a default embed. See: https://github.com/4dn-dcic/snovault/pull/97/files#diff-37e50a5b28e424b1f12603c8d6cd6e05R221
- **tests**: updated 'em

### CODE REVIEW

File: docs/embedding-and-indexing.rst [hunk 1]
```diff
@@ -45,9 +45,9 @@ The next embed, 'award.title', has a terminal field. This means we are only inte
     award.principals_allowed.*
 ]
 
-The last embed is 'submitted_by'. This will actually throw a warning from the tests written around fourfront_utils, since this embed is completely unnecessary. Since submitted_by is a top-level linkTo in our object, there is no reason to add it to the embedded_list; it will already get link_id, display_title, and uuid added. So let's change this embed to submitted_by.some_object (where some_object is a fictional linkTo). This would automatically add @id, principals_allowed, link_id, display_title, and uuid for some_object.
+The last embed is 'submitted_by'. This will actually throw a warning from the tests written around embedded_list, since this embed is completely unnecessary. Since submitted_by is a top-level linkTo in our object, there is no reason to add it to the embedded_list; it will already get link_id, display_title, and uuid added. So let's change this embed to submitted_by.some_object (where some_object is a fictional linkTo). This would automatically add @id, principals_allowed, link_id, display_title, and uuid for some_object.
```
Comment (Chadius) [inline docs/embedding-and-indexing.rst]: We're getting rid of link_id, right? So the updated text shouldn't mention the field.

File: src/snovault/resources.py [hunk 2]
```diff
@@ -461,3 +469,33 @@ def jsonld_type(self):
...
+        'properties': {
+            'view': {
+                'type': 'string'
+            },
+            'edit': {
+                'type': 'string'
+            },
+            'audit': {
+                'type': 'string'
+            }
```
Comment (Chadius) [inline src/snovault/resources.py]: principals_allowed - should this be an array of strings?

File: docs/embedding-and-indexing.rst (hunk 1 above)
Comment (carlvitzthum) [reply docs/embedding-and-indexing.rst]: Good catch, will make a tweak.

File: src/snovault/resources.py (hunk 2 above)
Comment (carlvitzthum) [reply src/snovault/resources.py]: Nope, it's a dictionary with different permissions for `view`, `edit`, and `audit` (the keys)

**Pull request merged**